### DISCUSSION TITLE
Panic when setting config value

### DISCUSSION THREAD

**Discussion by lhjt** [jdx/mise#3820] [Troubleshooting and bug reports]
**Describe the bug**
Trying to set a config value as [per the documentation](https://mise.jdx.dev/dev-tools/#tool-options) results in a panic:

```
$ mise config set tools.node.postinstall 'corepack enable'
The application panicked (crashed).
Message:  called `Option::unwrap()` on a `None` value
Location: src/cli/config/set.rs:102

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
```

**To Reproduce**

1. Install mise
2. Create a new project folder and run: `mise u node@lts`
3. Try and enable corepack via postinstall command: `mise config set tools.node.postinstall 'corepack enable'`
4. See panic

**Expected behavior**

Mise should not panic. If a tool has been installed where a key is set to a string indicating the version (such as `node = "lts"`) it should rewrite it automatically to the expanded output `node = { version = 'lts' }`

**`mise doctor` output**

```text
version: 2024.12.19 linux-x64 (82840ee 2024-12-23)
activated: yes
shims_on_path: no

build_info:
  Target: x86_64-unknown-linux-gnu
  Features: DEFAULT, NATIVE_TLS, OPENSSL
  Built: Mon, 23 Dec 2024 19:10:54 +0000
  Rust Version: rustc 1.83.0 (90b35a623 2024-11-26)
  Profile: release

shell:
  /usr/bin/zsh
  zsh 5.9 (x86_64-ubuntu-linux-gnu)

dirs:
  cache: ~/.cache/mise
  config: ~/.config/mise
  data: ~/.local/share/mise
  shims: ~/.local/share/mise/shims
  state: ~/.local/state/mise

config_files:
  ~/.config/mise/config.toml
  ~/repos/gpx-earth/mise.toml

ignored_config_files: (none)

backends:
  aqua
  asdf
  cargo
  core
  dotnet
  gem
  go
  npm
  pipx
  spm
  ubi
  vfox

plugins:

toolset:
  aqua:jqlang/jq@1.7.1
  core:node@22.12.0
  ubi:jdx/usage@1.7.4

path:
  ~/.local/share/mise/installs/node/22.12.0/bin
  ~/.local/share/mise/installs/jq/1.7.1
  ~/.local/share/mise/installs/usage/1.7.4/bin
  ~/.local/bin
  ~/.cargo/bin
  /usr/local/sbin
  /usr/local/bin
  /usr/sbin
  /usr/bin
  /sbin
  /bin
  /usr/games
  /usr/local/games
  /usr/lib/wsl/lib
  /opt/nvim-linux64/bin

env_vars:
  MISE_SHELL=zsh

settings:

No problems found
```

**Additional context**

```log
TRACE  1 [src/cli/mod.rs:327] logger done
TRACE  1 [src/cli/mod.rs:328] migrate start
TRACE  1 [src/cli/mod.rs:328] migrate done
DEBUG  1 [src/cli/mod.rs:333] ARGS: /home/user/.local/bin/mise config set tools.node.postinstall corepack enable --trace
TRACE  1 [src/cli/mod.rs:339] run config start
The application panicked (crashed).
Message:  called `Option::unwrap()` on a `None` value
Location: src/cli/config/set.rs:102

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
```

**Comment by roele** [comment]
It seems that the `config set` command cannot deal with this use case. This seems to be related to the fact that the TOML has simple tables (e.g. node = "latest") and deeper nested inline tables (e.g. node = { version = "latest" }) and the later are not (yet) taken into account.
From a user experience perspective it would make sense to deal with this automatically when having a `node = "latest"` to transform this automatically to a `node = { version = "latest" }` when adding other options.

I drafted a rough patch at https://github.com/jdx/mise/pull/3823

**Comment by noirbizarre** [comment]
It seems to also happen if the key exists elsewhere.

I have the same issue with:
```toml
[env]
_.python.venv = { path = ".venv", create = true }
```
when I execute:
``` shell
mise config set env._.python.venv.path .venv
```

It works if I change `mise.toml` for:
```toml
[env._.python.venv]
path = ".venv"
create = true
```

**Reply by jdx** [reply]
it appears @roele's patch will fix this when we get it

**Reply by roele** [reply]
It seems the fix will work for `path` but for `create` another issue appears as the type is not inferred correctly as types are only inferred for keys starting with `setting.`

**Comment by Marukome0743** [comment]
A follow-up to this has been merged in #11153.

The original panic (`mise config set tools.node.postinstall …` when `node = "lts"`) was fixed back in #3823 by converting a simple tool-version string into an inline table. #11153 handles the remaining related cases where a key segment collides with an existing **non-table (scalar)** value:

- Descending ≥2 levels through a scalar (e.g. `mise config set env.FOO.BAZ.QUX x` with `env.FOO = "bar"`) used to panic with `Option::unwrap() on a None value`.
- Setting a child of a scalar (e.g. `env.FOO.BAZ x`) used to silently do nothing (the write was discarded).

Both now fail with a clear `... is already set to a non-table value` error instead of panicking or silently no-op'ing. `mise config set` no longer clobbers an existing scalar into a table, so your data isn't lost. A case where the target is a pre-existing inline table (like @noirbizarre's `_.python.venv = { path = ".venv", create = true }`) already works and is now covered by a regression test.

Thanks @lhjt for the original report and @noirbizarre for spotting the non-table variant. The fix will be in the next release.

*This comment was generated by an AI coding assistant.*

**Reply by noirbizarre** [reply]
Thank you for the fix and for keeping us updated! 🙏🏻
